### PR TITLE
Add EFI_SHELL_PROTOCOL

### PR DESCRIPTION
**Pull request by Ayush1325**
Defined in UEFI Shell Specification.

I have left all the pointers as `*mut` for now. But I think it might be possible to change some of them to `*const`.

Signed-off-by: Ayush <ayushsingh1325@gmail.com>

### CODE REVIEW

**Comment by Ayush1325** [comment]
The failed CI is probably due to https://github.com/rust-lang/rust/pull/101088

**Comment by Ayush1325** [comment]
Hi @dvdhrm , can you rerun the CI now that https://github.com/rust-lang/rust/issues/101071 has been merged? I think it should be fixed now.

File: src/protocols/shell.rs [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+pub type GetMapFromDevicePath = eficall! {fn(
+    *mut *mut crate::protocols::device_path::Protocol,
+) -> *mut crate::base::Char16};
+
+pub type GetDevicePathFromFilePath = eficall! {fn(
+    *mut crate::base::Char16,
+) -> *mut crate::protocols::device_path::Protocol};
+
+pub type GetFilePathFromDevicePath = eficall! {fn(
+    *mut *mut crate::protocols::device_path::Protocol,
```
Comment (dvdhrm) [inline src/protocols/shell.rs:84]: This should be `*mut` rather than `*mut *mut`, right?

File: src/protocols/shell.rs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+pub const PROTOCOL_GUID: crate::base::Guid = crate::base::Guid::from_fields(
+    0x6302d008,
+    0x7f9b,
+    0x4f30,
+    0x87,
+    0xac,
+    &[0x60, 0xc9, 0xfe, 0xf5, 0xda, 0x4e],
+);
+
+pub type ShellFileHandle = *mut core::ffi::c_void;
```
Comment (dvdhrm) [inline src/protocols/shell.rs:14]: This is actually `SHELL_FILE_HANDLE` in the spec? Why does this not have the `EFI_*` prefix? *sigh*

If it was `EFI_SHELL_FILE_HANDLE`, we would call it `FileHandle` (drop the `shell` prefix). What do you think, should we rename to `FileHandle` to be more consistent?

File: src/protocols/shell.rs [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+pub const USE_DEVICE_PATH: DeviceNameFlags = 0x00000002;
+
+#[repr(C)]
+pub struct ListEntry {
+    pub flink: *mut ListEntry,
+    pub blink: *mut ListEntry,
+}
+
+#[repr(C)]
+pub struct ShellFileInfo {
```
Comment (dvdhrm) [inline src/protocols/shell.rs:27]: This is called `EFI_SHELL_FILE_INFO` in the spec, so we should really name it `FileInfo`.

**Review by dvdhrm — COMMENTED** [review]
Minor fixes that I have queued locally. Just one question before I merge it regarding `FileHandle`. See inline. I can do the fixup myself.

File: src/protocols/shell.rs (hunk 1 above)
Comment (Ayush1325) [reply src/protocols/shell.rs:84]: > This should be `*mut` rather than `*mut *mut`, right?

Yes. Should I fix that myself?

Comment (dvdhrm) [reply src/protocols/shell.rs:84]: No, I already have it queued locally.

File: src/protocols/shell.rs (hunk 2 above)
Comment (Ayush1325) [reply src/protocols/shell.rs:14]: I think it would be fine to rename it to `FileHandle` since it is not used in any other Shell Protocols.

**Comment by dvdhrm** [comment]
Perfect! Merged. Thanks a lot!

**Pull request closed without merging**